opacus/layers/dp_lstm.py

```diff
@@ -320,6 +431,8 @@ def forward(
             )
         else:
             h_0s = h_0s.reshape([L, P, B, H])
+            h_0s = self._permute_hidden(h_0s, sorted_indices, 2)
```

Comment:
Why do we need this?

Reply:
So, the PackedSequence always ensures that each of the "batches" within the 2 dimensional data tensor contain the encoding of the sequence in descending order of the sequence lengths even though the original data did not contain sequences in a sorted order(descending) of sequence length. Let me clarify with an example:
```python
from opacus.layers import DPLSTM
import torch 

seq_batch = [torch.tensor([[1, 1],
                           [2, 2],
                           [3, 3],
                           [4, 4],
                           [5, 5]]),
             torch.tensor([[10, 10],
                           [20, 20]])]

seq_lens = [5, 2]

padded_seq_batch = torch.nn.utils.rnn.pad_sequence(seq_batch, batch_first=True)
packed_seq_batch = torch.nn.utils.rnn.pack_padded_sequence(padded_seq_batch, lengths=seq_lens, batch_first=True)


seq_batch = [torch.tensor([[10, 10],
                           [20, 20]]),
            torch.tensor([[1, 1],
                           [2, 2],
                           [3, 3],
                           [4, 4],
                           [5, 5]])]

seq_lens = [2, 5]

padded_seq_batch = torch.nn.utils.rnn.pad_sequence(seq_batch, batch_first=True)
packed_seq_batch = torch.nn.utils.rnn.pack_padded_sequence(padded_seq_batch, lengths=seq_lens, batch_first=True, enforce_sorted=False)

```
Both approaches produce the same tensor within the PackedSequence. The ```_permute_hidden``` function makes sure that the arrangement of the hidden weights are in line with the data format within PackedSequence.

Reply:
Makes sense :) Thanks for explaining!